File: rhino/system/java/inonit/system/cygwin/Cygwin.java
//	LICENSE
//	This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
//	distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
//	The Original Code is the SLIME operating system interface.
//
//	The Initial Developer of the Original Code is David P. Caldwell <david@davidpcaldwell.com>.
//	Portions created by the Initial Developer are Copyright (C) 2010-2013 the Initial Developer. All Rights Reserved.
//
//	Contributor(s):
//	END LICENSE

package inonit.system.cygwin;

import java.io.*;

import inonit.script.runtime.io.*;
import inonit.system.*;

public class Cygwin {
	public static class CygpathException extends Exception {
		CygpathException(String message) {
			super(message);
		}
	}

	private static class CygpathCanonicalPathException extends RuntimeException {
		CygpathCanonicalPathException(IOException e) {
			super(e);
		}
	}

	private static String getCanonicalPath(File file) {
		try {
			return file.getCanonicalPath();
		} catch (IOException e) {
			throw new CygpathCanonicalPathException(e);
		}
	}

	private static String cygpath(String cygpath, String[] args) throws CygpathException {
		int TRIES = 10;
		String rv = "";
		int i = 0;
		while (rv.length() == 0 && i < TRIES) {
			Command.Result result = OperatingSystem.get().execute(cygpath, args);
			try {
				InputStream stream = result.evaluate().getOutputStream();
				Streams streams = new Streams();
				rv = streams.readString(stream);
			} catch (IOException e) {
				//	we are going to retry unless this is the TRIESth failure
			} catch (Command.Result.Failure e) {
				//	we are going to retry unless this is the TRIESth failure
			}
			if (rv.length() > 0) {
				rv = rv.substring(0,rv.length()-1);
			} else {
				i++;
			}
		}
		if (rv.length() == 0) {
			String message = "";
			for (int j=0; j<args.length; j++) {
				message += args[j];
				if (j+1 != args.length)
					message += ",";
			}
			throw new CygpathException("cygpath failed: cygpath=" + cygpath + " arguments=" + message);
		}
		return rv;
	}

	public static Cygwin locate() {
		try {
			return new Cygwin(new File(cygpath("cygpath", new String[] { "-w", "/" })));
		} catch (CygpathException e) {
			return null;
		}
	}

	public static Cygwin create(File root) {
		return new Cygwin(root);
	}

	private File root;

	Cygwin(File root) {
		this.root = root;
	}

	private String cygpath(String flags, String path) throws CygpathException {
		return cygpath(getCanonicalPath(new File(root, "bin/cygpath.exe")), new String[] { flags, path });
	}

	public String toWindowsPath(String name, boolean path) throws CygpathException {
		return cygpath((path) ? "-wp" : "-w", name);
	}

	public String toUnixPath(String name, boolean path) throws CygpathException {
		return cygpath((path) ? "-up" : "-u", name);
	}
}

File: rhino/file/cygwin.js
//	LICENSE
//	This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not
//	distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
//	The Original Code is the rhino/file SLIME module.
//
//	The Initial Developer of the Original Code is David P. Caldwell <david@davidpcaldwell.com>.
//	Portions created by the Initial Developer are Copyright (C) 2013 the Initial Developer. All Rights Reserved.
//
//	Contributor(s):
//	END LICENSE

var _cygwinProvider;
if ($context.cygwin.root && !$context.cygwin.paths) {
	_cygwinProvider = Packages.inonit.script.runtime.io.cygwin.CygwinFilesystem.create($context.cygwin.root)
} else {
	_cygwinProvider = Packages.inonit.script.runtime.io.cygwin.CygwinFilesystem.create($context.cygwin.root,$context.cygwin.paths)
}
var cygwinProvider = new $context.java.FilesystemProvider(_cygwinProvider);
$exports.filesystem = new $context.Filesystem(cygwinProvider, {
	interpretNativePathname: function(pathname) {
		return this.toUnix(pathname);
	}
});

$exports.filesystem.toUnix = function(item) {
	if (isPathname(item)) {
		return new $context.Pathname({ filesystem: cygwinProvider, peer: _cygwinProvider.getNode( item.java.adapt() ) });
	}
	if (item instanceof $context.Searchpath) {
		return new $context.Searchpath({ filesystem: cygwinProvider, array: item.pathnames });
	}
	return item;
}

$exports.filesystem.toWindows = function(item) {
	if ($context.isPathname(item)) {
		//	Unbelievably horrendous workaround, but seems to work
		//	When creating a softlink to an exe in Windows, the softlink gets the .exe suffix added to it even if it is not on the
		//	command line.
		if (item.file == null && this.Pathname( item.toString() + ".exe" ).file != null ) {
			item = this.Pathname( item.toString() + ".exe" );
		}
		return $context.java.FilesystemProvider.os.importPathname( item );
	}
	//	Searchpath currently sets the constructor property to this module-level function; would this make this instanceof
	//	work?
	if (item instanceof $context.Searchpath) {
		//	TODO	convert underlying pathnames
		return new $context.Searchpath({ filesystem: $context.java.FilesystemProvider.os, array: item.pathnames });
	}
	return item;
}

if ($context.addFinalizer) {
	$context.addFinalizer(function() {
		_cygwinProvider.finalize();
	});
}
